imdclient/IMDServer.py

```diff
@@ -0,0 +1,357 @@
+        **kwargs,
+    ):
+        if version not in IMDVERSIONS:
+            raise ValueError(
+                f"IMDServer: Incompatible IMD version. Expected version in {IMDVERSIONS}, got {version}"
+            )
+        if version == 2 and (time or box or velocities or forces):
+            # TODO: and energies?
+            raise ValueError("IMDServer: IMDv2 only supports positions")
+        if (version == 2 and not (positions)) or (
+            version == 3
+            and not (time or box or positions or velocities or forces)
+        ):
+            raise ValueError(
+                f"IMDServer: No IMD flags turned on for IMDv{version}"
+            )
+
+        end = ">" if sys.byteorder == "big" else "<"
+
+        self.sinfo = IMDSessionInfo(
+            version=version,
+            endianness=end,
+            wrapped_coords=False,
+            time=time,
+            energies=False,
+            box=box,
+            positions=positions,
+            velocities=velocities,
+            forces=forces,
+        )
+        self.n_atoms = n_atoms
+
+        # shutdown safety
+        signal.signal(signal.SIGINT, self.signal_handler)
+        signal.signal(signal.SIGTERM, self.signal_handler)
+        try:
+            import IPython
+        except ImportError:
+            has_ipython = False
+        else:
+            has_ipython = True
+
+        if has_ipython:
+            try:
+                from IPython import get_ipython
+
+                if get_ipython() is not None:
+                    kernel = get_ipython().kernel
+                    kernel.pre_handler_hook = lambda: None
+                    kernel.post_handler_hook = lambda: None
+                    logger.debug("Running in Jupyter")
+            except NameError:
+                logger.debug("Running in non-jupyter IPython environment")
+
+        atexit.register(self.stop)
+
+        self._alloc_send_buf()
+        self._header = bytearray(IMDHEADERSIZE)
+        self._array_dtype = np.dtype(f"{self.sinfo.endianness}f4")
+        self._time_fmt = f"{self.sinfo.endianness}ddQ"
+
+        try:
+            self._listen_socket = socket.socket(
+                socket.AF_INET, socket.SOCK_STREAM
+            )
+            self._listen_socket.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
+            self._listen_socket.bind((host, port))
+
+            # for the case of port=0
+            self.port = self._listen_socket.getsockname()[1]
+            self._listen_socket.listen()
+
+            if sock_contains_data(self._listen_socket, listen_timeout):
+                self._conn, _ = self._listen_socket.accept()
+            else:
+                # TODO: abstract into loop
+                raise TimeoutError(
+                    f"IMDServer: No client within {listen_timeout} seconds"
+                )
+
+            if version == 2:
+                self._send_handshakeV2()
+            elif version == 3:
+                self._send_handshakeV3()
```

Comment:
I know that you're checking `version` above. It still makes me squirm to have no `else` so I'd add `assert 2 <= version <= 3`, just in case the `IMDVERSIONS` get updated but nothing is done here.